migration/migrator_test.go

```diff
@@ -0,0 +1,287 @@
+package migration
+
+import (
+	"fmt"
+	"math/big"
+	"math/rand"
+	"testing"
+	"time"
+
+	zktrie "github.com/kroma-network/zktrie/types"
+	"github.com/stretchr/testify/require"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/consensus/ethash"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/rawdb"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/core/vm"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/ethereum/go-ethereum/trie"
+	"github.com/ethereum/go-ethereum/trie/trienode"
+)
+
+func TestApplyNewStateTransition(t *testing.T) {
+	genesis := &core.Genesis{
+		Config: params.KromaTestConfig,
+	}
+	chainDB := rawdb.NewMemoryDatabase()
+	cacheConfig := core.DefaultCacheConfigWithScheme(rawdb.HashScheme)
+	cacheConfig.Preimages = true
+	cacheConfig.KromaZKTrie = true
+	blockchain, _ := core.NewBlockChain(chainDB, cacheConfig, genesis, nil, ethash.NewFaker(), vm.Config{}, nil, nil)
+	blockchain.TrieDB().SetBackend(true)
+
+	eth := fakeEthBackend{
+		chainDB, blockchain,
+	}
+
+	m, err := NewStateMigrator(&eth)
+	require.NoError(t, err)
+
+	for i := 0; i < 50; i++ {
+		err = addBlockWithRandomChanges(m, []common.Address{
+			common.HexToAddress("0x0000000000000000000000000000000000001234"),
+			common.HexToAddress("0x0000000000000000000000000000000000001235"),
+			common.HexToAddress("0x0000000000000000000000000000000000001236"),
+		})
+		require.NoError(t, err)
+	}
+
+	head := rawdb.ReadHeadHeader(m.db)
+	err = m.migrateAccount(head)
+	require.NoError(t, err)
+
+	err = m.ValidateStateWithIterator(m.migratedRef.Root(), head.Root)
+	require.NoError(t, err)
+
+	accountsNum := 20
+	var addresses []common.Address
+	for i := 0; i < accountsNum; i++ {
+		addresses = append(addresses, common.BigToAddress(big.NewInt(int64(i))))
+	}
+
+	for i := 0; i < 50; i++ {
+		err = addBlockWithRandomChanges(m, addresses)
+		require.NoError(t, err)
+		head = rawdb.ReadHeadHeader(m.db)
+		err = m.applyNewStateTransition(head.Number.Uint64())
+		require.NoError(t, err)
+	}
+
+	destructChanges := make(map[common.Address]*types.StateAccount)
+
+	for i := 0; i < accountsNum; i += 2 {
+		destructChanges[addresses[i]] = nil
+	}
+	err = addBlock(m, destructChanges, nil, nil)
+	require.NoError(t, err)
+
+	head = rawdb.ReadHeadHeader(m.db)
+	err = m.applyNewStateTransition(head.Number.Uint64())
+	require.NoError(t, err)
+
+	err = addBlockWithRandomChanges(m, []common.Address{addresses[0]})
+	require.NoError(t, err)
+
+	head = rawdb.ReadHeadHeader(m.db)
+	err = m.applyNewStateTransition(head.Number.Uint64())
+	require.NoError(t, err)
```

Comment:
I think it would be more meaningful to apply stateTransition to MPT tries as well for the parts that currently only check for no error and compare if they match.